lib/memo_wise.rb

```diff
@@ -179,73 +180,68 @@ def #{method_name}
+          # `@_memo_wise_indices` stores the `@_memo_wise_single_argument`
+          # indices of different method names. We only use this data structure
+          # when resetting or presetting memoization. It looks like:
+          #   {
+          #     ClassBeingMemoized => {
+          #       single_arg_method_name: 0,
+          #       other_single_arg_method_name: 1
+          #     },
+          #     ClassBeingMemoizedSingletonClass => {
+          #       another_single_arg_method_name: 0
+          #     }
+          #   }
+          @_memo_wise_indices ||= Hash.new { |h, k| h[k] = {} }
+          @_memo_wise_index_counters ||= Hash.new(0)
+
+          index = @_memo_wise_index_counters[klass]
+          @_memo_wise_indices[klass][method_name] = index
+          @_memo_wise_index_counters[klass] += 1
+
+          key = method.parameters.first.last
+
+          klass.module_eval <<-END_OF_METHOD, __FILE__, __LINE__ + 1
+            def #{method_name}(#{MemoWise::InternalAPI.args_str(method)})
+              hash = (@_memo_wise_single_argument[#{index}] ||= {})
+              output = hash[#{key}]
+              if output || hash.key?(#{key})
+                output
+              else
+                hash[#{key}] = #{original_memo_wised_name}(#{MemoWise::InternalAPI.call_str(method)})
+              end
             end
-          else
-            # If our method has arguments, we need to separate out our handling
-            # of normal args vs. keyword args due to the changes in Ruby 3.
-            # See: <link>
-            # By only including logic for *args, **kwargs when they are used in
-            # the method, we can avoid allocating unnecessary arrays and hashes.
-            has_arg = MemoWise::InternalAPI.has_arg?(method)
-
-            if has_arg && MemoWise::InternalAPI.has_kwarg?(method)
-              args_str = "(*args, **kwargs)"
-              fetch_key_init = "[:#{method_name}, args, kwargs].hash"
-              use_hashed_key = true
-            elsif has_arg
-              args_str = "(*args)"
-              fetch_key_init = "args.hash"
-            else
-              args_str = "(**kwargs)"
-              fetch_key_init = "kwargs.hash"
+          END_OF_METHOD
+        when :multiple_required, :splat_and_double_splat
+          klass.module_eval <<-END_OF_METHOD, __FILE__, __LINE__ + 1
+            def #{method_name}(#{MemoWise::InternalAPI.args_str(method)})
+              key = #{MemoWise::InternalAPI.key_str(method)}
+              output = @_memo_wise[key]
+              if output || @_memo_wise.key?(key)
+                output
+              else
+                hashes = (@_memo_wise_hashes[:#{method_name}] ||= Set.new)
+                hashes << key
+                @_memo_wise[key] = #{original_memo_wised_name}(#{MemoWise::InternalAPI.call_str(method)})
+              end
             end
-          end
+          END_OF_METHOD
+        else # :splat, :double_splat
```

Comment:
Because then we need to add a test that exercises the `else` case, which is impossible/very difficult because our cases are exhaustive. Does that make sense?